irohad/model/impl/query_execution.cpp

```diff
@@ -76,374 +90,344 @@ bool hasQueryPermission(const std::string &creator,
-    }
-    return executeGetSignatories(*qry);
-  }
-  if (instanceof <GetAccountTransactions>(query.get())) {
-    auto qry = std::static_pointer_cast<const GetAccountTransactions>(query);
-    if (!validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = model::ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetAccountTransactions(*qry);
-  }
-  if (instanceof <GetAccountAssetTransactions>(query.get())) {
-    auto qry =
-        std::static_pointer_cast<const GetAccountAssetTransactions>(query);
-    if (!validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = model::ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetAccountAssetTransactions(*qry);
-  }
-  if (instanceof <GetTransactions>(query.get())) {
-    auto qry = std::static_pointer_cast<const GetTransactions>(query);
-    if (not validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetTransactions(*qry);
-  }
-  if (instanceof <GetRoles>(query.get())) {
-    auto qry = std::static_pointer_cast<const GetRoles>(query);
-    if (not validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetRoles(*qry);
-  }
-  if (instanceof <GetRolePermissions>(query.get())) {
-    auto qry = std::static_pointer_cast<const GetRolePermissions>(query);
-    if (not validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetRolePermissions(*qry);
-  }
-  if (instanceof <GetAssetInfo>(query.get())) {
-    auto qry = std::static_pointer_cast<const GetAssetInfo>(query);
-    if (not validate(*qry)) {
-      ErrorResponse response;
-      response.query_hash = iroha::hash(*qry);
-      response.reason = ErrorResponse::STATEFUL_INVALID;
-      return std::make_shared<ErrorResponse>(response);
-    }
-    return executeGetAssetInfo(*qry);
-  }
-  ErrorResponse response;
-  response.query_hash = iroha::hash(*query);
-  response.reason = model::ErrorResponse::NOT_SUPPORTED;
-  return std::make_shared<ErrorResponse>(response);
+std::shared_ptr<shared_model::interface::QueryResponse>
+QueryProcessingFactory::execute(const shared_model::interface::Query &query) {
+  const auto &query_hash = query.hash();
+  QueryResponseBuilderDone builder;
+  return visit_in_place(
```

Comment:
Please add TODO with task id to rewrite this as in https://github.com/hyperledger/iroha/pull/1007/files#diff-cdb3f6acf132ab4b63459a816a9a50b5R80 , so that code is not duplicated.